app/assets/javascripts/discourse/app/lib/plugin-api.js

```diff
@@ -1648,10 +1651,37 @@ class PluginApi {
   addSaveableUserOptionField(fieldName) {
     addSaveableUserOptionField(fieldName);
   }
+
+  /**
+   * Adds additional params to be sent to the reviewable/:id/perform/:action
+   * endpoint for a given reviewable type. This is so plugins can provide more
+   * complex reviewable actions that may depend on a custom modal.
+   *
+   * This is copied from the reviewable model instance when performing an action
+   * on the ReviewableItem component.
+   *
+   * ```
+   * api.addPluginReviewableParam("ReviewablePluginType", "some_param");
+   * ```
+   **/
   addPluginReviewableParam(reviewableType, param) {
     addPluginReviewableParam(reviewableType, param);
   }
 
+  /**
+   * Registers a modal JS class for a specific reviewable type. This is used
+   * alongside the `custom_modal` option for reviewables. The `custom_modal`
+   * option should just be the JS class name as a string, then we map it to the
+   * real class at runtime.
+   *
+   * ```
+   * api.registerReviewableCustomModal("ReviewablePluginType", MyCustomModal);
```

Comment:
I think we need to expand on how `MyCustomModal` should be implemented because `modal JS class` is not very descriptive of what functions the class needs to implement. Also, is there a reason we need to pass it in as a string instead of just passing the actual class directly?

Reply:
@tgxworld added some better docs.

> Also, is there a reason we need to pass it in as a string instead of just passing the actual class directly?

Yeah the first param needs to be a string because it references a ruby class...the second param is an actual JS class. Then on the server-side it flips, you can see https://github.com/discourse/discourse-category-experts/pull/117/files for reference

Reply:
I think the fact we have to pass in a Ruby class here is a code smell. Client side API should technically not have to be concerned about the server side. I'm don't have too much context here but is there a way we can introduce an API that is independent of the server side code?

Reply:
@tgxworld I just went with this because we already have an API for reviewables that uses a server class name as the first argument:

```js
api.addPluginReviewableParam(
  "ReviewableCategoryExpertSuggestion",
  "group_id"
);
```

We basically need some way to restrict/namespace these JS classes to a specific reviewable type. I'm not sure what other way we can do this, unless we just directly map the JS string class to the real class like this:

```javascript
api.registerReviewableCustomModal("MyCustomModal", MyCustomModal);
```

And just hope that it doesn't class between plugins (which I think is unlikely anyway).